core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```
@@ -0,0 +1,165 @@
+     */
+    public void cancel() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      // TODO(calvin): Log a warning if the delete fails
+      ufs.delete(mTemporaryPath, false);
+    }
+
+    /**
+     * Closes the temporary file and attempts to promote it to the final file path. If the final
+     * path already exists, the stream is canceled instead.
+     *
+     * @throws IOException if an error occurs during the under file system operation
+     */
+    public void close() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (!ufs.rename(mTemporaryPath, mPath)) {
+        // TODO(calvin): Log a warning if the delete fails
+        ufs.delete(mTemporaryPath, false);
+      }
+    }
+  }
+
+  /** A random id generator for worker file ids. */
+  private AtomicLong mIdGenerator;
+  /** Map of worker file ids to open under file system streams. */
+  private ConcurrentMap<Long, UnderFileSystemOutputStream> mStreams;
+
+  /**
+   * Creates a new under file system manager. Stream ids are unique to each under file system
+   * manager.
+   */
+  public UnderFileSystemManager() {
+    mIdGenerator = new AtomicLong(new Random().nextLong());
+    mStreams = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Creates a {@link UnderFileSystemOutputStream} for the given file path and adds it to the
+   * open streams map keyed with the returned worker file id.
+   *
+   * @param ufsPath the path to create in the under file system
+   * @return the worker file id which should be used to reference the open stream
+   * @throws FileAlreadyExistsException if the under file system path already exists
+   * @throws IOException if an error occurs when operating on the under file system
+   */
+  public long createFile(String ufsPath) throws FileAlreadyExistsException, IOException {
+    long workerFileId = mIdGenerator.getAndIncrement();
```

generate the id after the stream creation to fail fast?